File: Core/src/main/java/jokatu/components/stomp/StoringMessageSender.java
package jokatu.components.stomp;

import jokatu.components.stores.MessageStorer;
import org.jetbrains.annotations.NotNull;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.messaging.Message;
import org.springframework.messaging.MessagingException;
import org.springframework.messaging.simp.SimpMessagingTemplate;
import org.springframework.messaging.support.GenericMessage;
import org.springframework.stereotype.Component;

/**
 * Sends messages to STOMP destinations and stores a history of them.
 */
@Component
public class StoringMessageSender {

	private final MessageStorer messageStore;
	private final SimpMessagingTemplate simpMessagingTemplate;

	@Autowired
	public StoringMessageSender(MessageStorer messageStore, SimpMessagingTemplate simpMessagingTemplate) {
		this.messageStore = messageStore;
		this.simpMessagingTemplate = simpMessagingTemplate;
	}

	public void sendToUser(@NotNull String user, @NotNull String destination, @NotNull Object payload) throws MessagingException {
		sendMessageToUser(user, destination, new GenericMessage<>(payload));
	}

	public void sendMessageToUser(@NotNull String user, @NotNull String destination, @NotNull Message message) throws MessagingException {
		messageStore.storeForUser(user, destination, message);
		simpMessagingTemplate.convertAndSendToUser(user, destination, message.getPayload(), message.getHeaders());
	}

	public void send(@NotNull String destination, @NotNull Object payload) throws MessagingException {
		sendMessage(destination, new GenericMessage<>(payload));
	}

	public void sendMessage(@NotNull String destination, @NotNull Message message) throws MessagingException {
		messageStore.store(destination, message);
		simpMessagingTemplate.convertAndSend(destination, message.getPayload(), message.getHeaders());
	}
}

File: Core/src/main/kotlin/jokatu/components/stomp/StoringMessageSender.kt
package jokatu.components.stomp

import jokatu.components.stores.MessageStorer
import org.slf4j.getLogger
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.messaging.Message
import org.springframework.messaging.MessagingException
import org.springframework.messaging.simp.SimpMessagingTemplate
import org.springframework.messaging.support.GenericMessage
import org.springframework.stereotype.Component

/**
 * Sends messages to STOMP destinations and stores a history of them.
 */
@Component
class StoringMessageSender @Autowired constructor(
		private val messageStore: MessageStorer,
		private val simpMessagingTemplate: SimpMessagingTemplate
) {

	@Throws(MessagingException::class)
	fun <T: Any> sendToUser(user: String, destination: Destination<T>, payload: T) {
		sendMessageToUser(user, destination, GenericMessage(payload))
	}

	@Throws(MessagingException::class)
	fun <T: Any> sendMessageToUser(user: String, destination: Destination<T>, message: Message<T>) {
		log.trace("Sending message '$message' to $destination for $user")
		messageStore.storeForUser(user, destination, message)
		simpMessagingTemplate.convertAndSendToUser(user, destination.toString(), message.payload, message.headers)
	}

	@Throws(MessagingException::class)
	fun <T: Any> send(destination: Destination<T>, payload: T) {
		sendMessage(destination, GenericMessage(payload))
	}

	@Throws(MessagingException::class)
	fun <T: Any> sendMessage(destination: Destination<T>, message: Message<T>) {
		log.trace("Sending message '$message' to $destination")
		messageStore.store(destination, message)
		simpMessagingTemplate.convertAndSend(destination.toString(), message.payload, message.headers)
	}

	companion object {
		private val log = getLogger(StoringMessageSender::class)
	}
}
